### PR TITLE
Update amp-form docs and add an expanded note on CORS security.

### PR DESCRIPTION
**Pull request by mkhatib**
ITI: #3343

### CODE REVIEW

**Comment by erwinmombay** [comment]
adding @dvoytenko to this since he has done a few of the CORS documentation

File: extensions/amp-form/amp-form.md [hunk 1]
```diff
@@ -75,7 +75,108 @@ __required__
 
 Action must be provided, `https` and is non-cdn link (does **NOT** link to https://cdn.ampproject.org).
 
+**action-xhr**
+__optional__
+You can also provide an action-xhr attribute, if provided, the form will be submitted in an XHR fashion.
+
+This attribute can be the same or a different endpoint that action and has the same action requirement above.
```
Comment (dvoytenko) [inline extensions/amp-form/amp-form.md]: "than action" and let's make it `action`.

Comment (dvoytenko) [inline extensions/amp-form/amp-form.md]: Also, "requirements" (plural)

File: extensions/amp-form/amp-form.md [hunk 2]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+#### Invalid Submit Blocking and Validation Message Bubble
+Browsers that uses webkit-based engines, currently (as of August 2016) do not support invalid form submissions. These include Safari on all platforms, and all iOS browsers. `amp-form` polyfill this behavior to block any invalid submissions and show validation message bubbles on invalid inputs.
+
+**Note**: Messages are sometimes limited to few words like "required field", these messages are provided by the browser implementation. We'll be working on allowing publisher-provided custom validation messages as well as custom validation UIs (instead of the builtin/polyfill'd bubbles).
+
+#### :user-invalid/:user-valid
+These pseudo classes are part of [future CSS Selectors 4 spec](https://drafts.csswg.org/selectors-4/#user-pseudos) and are introduced to allow better hooks for styling invalid/valid fields based on few criteria.
+
+One of the main differences between `:invalid` and `:user-invalid` is when are they applied to the element. The latter being focused on being applied after a significant interaction from the user with the field (e.g. User types in a field, or blur from the field).
```
Comment (dvoytenko) [inline extensions/amp-form/amp-form.md]: "The latter being focused" - let's rewrite as "`:user-invalid` is applied after significant interaction..."

**Comment by dvoytenko** [comment]
LGTM

**Comment by dvoytenko** [comment]
LGTM with few comments

File: extensions/amp-form/amp-form.md [hunk 3]
```diff
@@ -75,7 +75,108 @@ __required__
...
+**Important**: Your XHR endpoints need to follow and implement [CORS Requests in AMP spec](https://github.com/ampproject/amphtml/blob/master/spec/amp-cors-requests.md).
+
 All other [form attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) are optional.
 
 ## Inputs
 Currently, `<input type=button>`, `<input type=file>`, `<input type=image>` and `<input type=password>` are not allowed. (This might be reconsidered in the future - please let us know if you require these and use cases).
+
+## Events
+`amp-form` exposes 3 events:
+**submit**
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: Can you add a line before this line. The "submit" line is showing up at the end of the above line with the colon.

Might also want to consider making these three a bulleted list.

File: extensions/amp-form/amp-form.md [hunk 4]
```diff
@@ -75,7 +75,108 @@ __required__
...
+Emitted whenever the form is submitted and before the submission is complete.
+
+**submit-success**
+Emitted whenever the form submission is done and response is a success.
+
+**submit-error**
+Emitted whenever the form submission is done and response is an error.
+
+These events can be used through the [on attribute](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-format.md#on).
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: `on`

File: extensions/amp-form/amp-form.md [hunk 5]
```diff
@@ -75,7 +75,108 @@ __required__
...
+For example, the following listens to both submit success and submit error and shows different lightboxes depending on the event
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: `submit-success` and `submit-error`.

Add period at end of sentence.

File: extensions/amp-form/amp-form.md [hunk 6]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+**Error Response**
+```json
+{
+  "name": "Jane Miller",
+  "message": "The email (email@example.com) you used is already subscribed."
+}
+```
+
+Publishers can render these in a template inside their forms as follow.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: as follows (plural)

File: extensions/amp-form/amp-form.md [hunk 7]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+```html
+<form ...>
+    <fieldset>
+      ...
+    </fieldset>
+    <div submit-success>
+        <template type="amp-mustache">
+            Success! Thanks {{name}} for subscribing! Please make sure to check your email {{email}}
+            to confirm! After that we'll start sending you weekly articles on {{#interests}}<b>{{name}}</b> {{{{/interests}}.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: is this syntax correct? it looks like the braces are unbalanced

File: extensions/amp-form/amp-form.md [hunk 8]
```diff
@@ -75,7 +75,108 @@ __required__
...
+            Oops! {{name}}, {{message}}.
+        </template>
+    </div>
+</form>
+```
+
+See the [full example here](https://github.com/ampproject/amphtml/blob/master/examples/forms.amp.html).
+
+## Polyfills
+`amp-form` polyfill behaviors and functionality missing from some browsers or being implemented in the next version of CSS.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: "polyfills" or "provides polyfills for"

File: extensions/amp-form/amp-form.md [hunk 9]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+#### Invalid Submit Blocking and Validation Message Bubble
+Browsers that uses webkit-based engines, currently (as of August 2016) do not support invalid form submissions. These include Safari on all platforms, and all iOS browsers. `amp-form` polyfill this behavior to block any invalid submissions and show validation message bubbles on invalid inputs.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: no comma

Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: "`amp-form` polyfills ..."

File: extensions/amp-form/amp-form.md [hunk 10]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+**Note**: Messages are sometimes limited to few words like "required field", these messages are provided by the browser implementation. We'll be working on allowing publisher-provided custom validation messages as well as custom validation UIs (instead of the builtin/polyfill'd bubbles).
+
+#### :user-invalid/:user-valid
+These pseudo classes are part of [future CSS Selectors 4 spec](https://drafts.csswg.org/selectors-4/#user-pseudos) and are introduced to allow better hooks for styling invalid/valid fields based on few criteria.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: "part of the"

Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: "based on a few criteria" ?

also, describe the criteria?

File: extensions/amp-form/amp-form.md (hunk 2 above)
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: lowercase "user" after e.g.

File: extensions/amp-form/amp-form.md [hunk 11]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+One of the main differences between `:invalid` and `:user-invalid` is when are they applied to the element. The latter being focused on being applied after a significant interaction from the user with the field (e.g. User types in a field, or blur from the field).
+
+`amp-form` provides classes (See below) to polyfill these pseudo-classes. `amp-form` also propagates these to ancestors `fieldset`s and `form`.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: lowercase "see"

File: extensions/amp-form/amp-form.md [hunk 12]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+
+## Classes and CSS Hooks
+`amp-form` provides few classes and CSS hooks for publishers to style their forms and inputs.
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: just "provides classes and CSS hooks"

File: extensions/amp-form/amp-form.md [hunk 13]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+`.amp-form-submitting`, `.amp-form-submit-success` and `.amp-form-submit-error` are added to indicate the state of the form submission.
+
+`.user-valid` and `.user-invalid` classes are a polyfill for the pseudo classes as described above. Publishers can use these to style their inputs and fieldsets to be more responsive with the user actions (e.g. highlighting an invalid input after user blurs from it with a red border).
```
Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: "to be responsive to user actions"

Comment (rudygalfi) [inline extensions/amp-form/amp-form.md]: "highlighting an invalid input with a red border after the user blurs from it"

File: spec/amp-cors-requests.md [hunk 14]
```diff
@@ -49,3 +49,11 @@ The resulting HTTP response has to also contain the following headers:
  - `Access-Control-Allow-Origin: <origin>`. Here "origin" refers to the requesting origin that was allowed via "Origin" request header above. Ex: "https://cdn.ampproject.org". This is a CORS spec requirement. Notice that while CORS spec allows the value of '\*' to be returned in this header, AMP strongly discourages use of '\*'. Instead the value of the "Origin" header should be validated and echoed for improved security.
  - `AMP-Access-Control-Allow-Source-Origin: <source-origin>`. Here "source-origin" indicates the source origin that is allowed to read the authorization response as was verified via "__amp_source_origin" URL parameter. Ex: "https://publisher1.com".
  - `Access-Control-Expose-Headers: AMP-Access-Control-Allow-Source-Origin`. This header simply allows CORS response to contain the "AMP-Access-Control-Allow-Source-Origin" header.
+
+#### Note on non-idempotent Requests
+When making CORS requests that would change the state of your system (e.g. User subscribes to or unsubscribes from a mailing list) the first two steps you need to make do are:
```
Comment (rudygalfi) [inline spec/amp-cors-requests.md]: lower case "user" after "e.g."

"you need to make sure to do"

File: spec/amp-cors-requests.md [hunk 15]
```diff
@@ -49,3 +49,11 @@ The resulting HTTP response has to also contain the following headers:
...
+
+1. Check the `Origin` header. If the origin was not `*.ampproject.org` or the publisher's origin, stop and return an error response.
```
Comment (rudygalfi) [inline spec/amp-cors-requests.md]: Can you explain more why you want to do each of these things? For instance, with this one, is the point that the publisher should only process stuff from trusted places? How do we future-proof on additional caches being trusted?

File: extensions/amp-form/amp-form.md (hunk 10 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: I linked to the specs which includes a better description for it and I mentioned below the user interaction criteria.

File: spec/amp-cors-requests.md (hunk 15 above)
Comment (mkhatib) [inline spec/amp-cors-requests.md]: I mentioned at the end that these provide  protection against CSRF attacks and added a small note on avoiding processing untrusted sources requests.

> How do we future-proof on additional caches being trusted?

Usually publishers would implement this in a central way (like in a middleware) in their backends, so adding more origins shouldn't be too troublesome. But this is a good point to discuss, maybe in another issue or next meeting?

File: extensions/amp-form/amp-form.md (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: done

File: extensions/amp-form/amp-form.md (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done.

File: extensions/amp-form/amp-form.md (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 7 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 8 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 9 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md [hunk 16]
```diff
@@ -75,7 +75,108 @@ __required__
 
 Action must be provided, `https` and is non-cdn link (does **NOT** link to https://cdn.ampproject.org).
 
+**action-xhr**
+__optional__
```
Comment (erwinmombay) [inline extensions/amp-form/amp-form.md]: add parentheses on __optional__

File: extensions/amp-form/amp-form.md (hunk 9 above)
Comment (erwinmombay) [inline extensions/amp-form/amp-form.md]: browsers that use

File: extensions/amp-form/amp-form.md [hunk 17]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+See the [full example here](https://github.com/ampproject/amphtml/blob/master/examples/forms.amp.html).
+
+## Polyfills
+`amp-form` polyfill behaviors and functionality missing from some browsers or being implemented in the next version of CSS.
+
+#### Invalid Submit Blocking and Validation Message Bubble
+Browsers that uses webkit-based engines, currently (as of August 2016) do not support invalid form submissions. These include Safari on all platforms, and all iOS browsers. `amp-form` polyfill this behavior to block any invalid submissions and show validation message bubbles on invalid inputs.
+
+**Note**: Messages are sometimes limited to few words like "required field", these messages are provided by the browser implementation. We'll be working on allowing publisher-provided custom validation messages as well as custom validation UIs (instead of the builtin/polyfill'd bubbles).
```
Comment (erwinmombay) [inline extensions/amp-form/amp-form.md]: to a few words

File: extensions/amp-form/amp-form.md (hunk 10 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 11 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 12 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 13 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: spec/amp-cors-requests.md (hunk 14 above)
Comment (mkhatib) [inline spec/amp-cors-requests.md]: Done

File: spec/amp-cors-requests.md (hunk 15 above)
Comment (mkhatib) [inline spec/amp-cors-requests.md]: Done

File: spec/amp-cors-requests.md [hunk 18]
```diff
@@ -49,3 +49,11 @@ The resulting HTTP response has to also contain the following headers:
...
+2. Check the `__amp_source_origin` query parameter. If it's not publisher's origin stop and return an error response.
```
Comment (erwinmombay) [inline spec/amp-cors-requests.md]: if its not the publisher's

File: extensions/amp-form/amp-form.md [hunk 19]
```diff
@@ -75,7 +75,108 @@ __required__
...
+
+**Important**: Your XHR endpoints need to follow and implement [CORS Requests in AMP spec](https://github.com/ampproject/amphtml/blob/master/spec/amp-cors-requests.md).
+
 All other [form attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) are optional.
 
 ## Inputs
 Currently, `<input type=button>`, `<input type=file>`, `<input type=image>` and `<input type=password>` are not allowed. (This might be reconsidered in the future - please let us know if you require these and use cases).
+
+## Events
+`amp-form` exposes 3 events:
```
Comment (erwinmombay) [inline extensions/amp-form/amp-form.md]: this is formatted wrong, theres no new line before **submit**

Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Updated in subsequent pushes.

Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: extensions/amp-form/amp-form.md (hunk 16 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done.

File: extensions/amp-form/amp-form.md (hunk 17 above)
Comment (mkhatib) [inline extensions/amp-form/amp-form.md]: Done

File: spec/amp-cors-requests.md (hunk 18 above)
Comment (mkhatib) [inline spec/amp-cors-requests.md]: Done

File: extensions/amp-form/amp-form.md (hunk 7 above)
Comment (erwinmombay) [inline extensions/amp-form/amp-form.md]: can you add a new line here so that the there are no horizontal scrollbars on the example to see the full code

**Pull request merged**